src/LtiAssignmentsGradesService.php

```diff
@@ -14,24 +14,45 @@ public function getScope(): array
         return $this->getServiceData()['scope'];
     }
 
+    // https://www.imsglobal.org/spec/lti-ags/v2p0#assignment-and-grade-service-claim
+    // When an LTI message is launching a resource associated to one and only one lineitem,
+    // the claim must include the endpoint URL for accessing the associated line item;
+    // in all other cases, this property must be either blank or not included in the claim.
+    public function getResourceLaunchLineItem(): ?LtiLineitem
+    {
+        $serviceData = $this->getServiceData();
+        if (empty($serviceData['lineitem'])) {
+            return null;
+        }
+
+        return LtiLineitem::new()->setId($serviceData['lineitem']);
+    }
+
     public function putGrade(LtiGrade $grade, LtiLineitem $lineitem = null)
     {
         if (!in_array(LtiConstants::AGS_SCOPE_SCORE, $this->getScope())) {
             throw new LtiException('Missing required scope', 1);
         }
+
+        // If the line item already contains the Id, we do not need to find or create it.
         if ($lineitem !== null && empty($lineitem->getId())) {
             $lineitem = $this->findOrCreateLineitem($lineitem);
-            $scoreUrl = $lineitem->getId();
-        } elseif ($lineitem === null && !empty($this->getServiceData()['lineitem'])) {
-            $scoreUrl = $this->getServiceData()['lineitem'];
-        } else {
-            $lineitem = LtiLineitem::new()
-                ->setLabel('default')
-                ->setScoreMaximum(100);
-            $lineitem = $this->findOrCreateLineitem($lineitem);
-            $scoreUrl = $lineitem->getId();
         }
 
+        // Otherwise, if no line item is passed in, attempt to use the one associated with
+        // this launch. If none exists, create a default line item.
+        else {
+            $lineitem = $this->getResourceLaunchLineItem();
+            if (!$lineitem || empty($lineitem->getId())) {
```

Comment:
Why do we also check for `empty($lineitem->getId())`? If `$this->getResourceLaunchLineItem()` returns `null`, this condition will be true, if `$this->getResourceLaunchLineItem()` doesn't return `null`, it seems that we will be setting the id there (line 28: `LtiLineitem::new()->setId($serviceData['lineitem']);`) Is there a chance we could set a `null` id?

Reply:
Hmm... I think you're right. @mattwright since you originally wrote this code, does this seem accurate to you?

Reply:
@zlayaAvocado is correct that we do not need to check empty($lineitem->getId()) as it will be set by $this->getResourceLaunchLineItem()

However, it looks like this branch is not using my code, but maybe a change to my code that you made? My code for putGrade was:

```php
        // If the line item already contains the Id, we do not need to find or create it.
        if ($lineitem !== null) {
            if (empty($lineitem->getId())) {
                $lineitem = $this->findOrCreateLineitem($lineitem);
            }
        }

        // Otherwise, if no line item is passed in, attempt to use the one associated with
        // this launch. If none exists, create a default line item.
        else {
            $lineitem = $this->getResourceLaunchLineItem();
            if (!$lineitem || empty($lineitem->getId())) {
                $lineitem = LtiLineitem::new()
                    ->setLabel('default')
                    ->setScoreMaximum(100);
                $lineitem = $this->findOrCreateLineitem($lineitem);
            }
        }
```

You've combined the first two conditional ifs into a single statement, which causes a fall-through to the else if they are not both true. but what we are really trying to do is:

1) If $lineitem is not null, then we keep it. However, inside that not-null check, if it doesn't have an id(), then we need to find or create it. This is why you cannot combine this into a single if() with two conditions, they need to remain separate. Does this make sense?

2) Else, If $lineitem is null, then, we check if there was a single line-item contained in the LTI launch. If so, we use that line item. Otherwise, we create the default one.

Reply:
And also, if $lineitem is not null when passed in, and already has an Id(), that is where we avoid some round trips to the LTI Provider so we don't need to "find" the lineitem we already have an id for.

Reply:
I refactored this out into `ensureLineItemExists()`. The logic is inverted from the previous implementation, but I think it reads more logically. I think I've properly accounted for the various edge cases, but let me know.